### ISSUE TITLE
Rule proposal: prefer readonly array

### ISSUE THREAD

**Issue by gajus** [typescript-eslint/typescript-eslint#8310]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

I would like an option that suggests that every array declaration is readonly.

### Fail Cases

```TypeScript
const items: Item[] = [];

const foo = (items: Item[]) => {};

const foo = (seed: { items: Item[] }) => {};
```

### Pass Cases

```TypeScript
const items: readonly Item[] = [];

const foo = (items: readonly Item[]) => {};

const foo = (seed: { items: readonly Item[] }) => {};
```

### Additional Info

Mixing readonly and non-readonly throughout the codebase can cause quite a headache.

Since mutating arrays is quite common, I'd like there to be an option that forces consistent use of readonly arrays.

**Comment by bradzacher** [comment]
Check out this FAQ article
https://typescript-eslint.io/linting/troubleshooting#how-can-i-ban-specific-language-feature
We no longer look to build rules that just ban syntax unless there's some real value we can add with a rule.

`readonly` is a `TSTypeOperator` node and the array type is `TSArrayType`. Using the `:not` selector you should be able to match all the types you care about, eg

```
:not(TSTypeOperator[operator = "readonly"]) > TSArrayType
```

Above is untested - you can test it in our playground (the AST tab has a query selector tester).

**Comment by JoshuaKGoldberg** [comment]
https://github.com/eslint-functional/eslint-plugin-functional would be a good place to ask about more targeted readonly rules too.
